File: todo/src/main/java/com/example/anvil/todo/Todo.java
package com.example.anvil.todo;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class Todo {
	private final static Todo sInstance = new Todo();

	private Todo() {}

	public static Todo getInstance() {
		return sInstance;
	}

	private List<TodoItem> items = new ArrayList<>();

	public List<TodoItem> items() {
		return items;
	}

	public void add(String msg) {
		items.add(new TodoItem(msg));
	}

	public void toggle(int index) {
		items.get(index).toggle();
	}

	public void clear() {
		Iterator iter = items.iterator();
		while (iter.hasNext()) {
			if (((TodoItem) iter.next()).isChecked()) {
				iter.remove();
			}
		}
	}

	public boolean hasChecked() {
		for (TodoItem item : items()) {
			if (item.isChecked()) {
				return true;
			}
		}
		return false;
	}

	public class TodoItem {
		private String message;
		private boolean checked;

		TodoItem(String msg) {
			message = msg;
		}

		public String getMessage() {
			return message;
		}

		public boolean isChecked() {
			return checked;
		}

		void toggle() {
			checked = !checked;
		}
	}
}

File: todo-kotlin/src/main/kotlin/com/example/anvil/todo/Todo.kt
package com.example.anvil.todo

object Todo {
	val items = mutableListOf<TodoItem>()

	fun add(msg: String) {
		items.add(TodoItem(msg, false))
	}

	fun toggle(index: Int) {
		val item = items[index]
		item.checked = !item.checked
	}

	fun clear() {
		items.filter { it.checked }.forEach { todoItem -> items.remove(todoItem) }
	}

	fun hasChecked() : Boolean = items.any { it.checked }

	data class TodoItem(val message: String, var checked: Boolean)
}



